Link time error with Apple Clang 17 and gfortran 14: ld: -lto_library library filename must be 'libLTO.dylib'

This morning, my Mac updated to the latest XCode Commandline Tools, aka Apple Clang 17:
```
❯ clang --version
Apple clang version 17.0.0 (clang-1700.0.13.3)
Target: arm64-apple-darwin24.4.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

I then nuked my Spack installs and rebuild my stack and when I did so the compilation died in OpenBLAS:
```
ld: -lto_library library filename must be 'libLTO.dylib'
collect2: error: ld returned 1 exit status
make[1]: *** [Makefile:320: sblat1] Error 1
make[1]: *** Waiting for unfinished jobs....
ld: -lto_library library filename must be 'libLTO.dylib'
collect2: error: ld returned 1 exit status
make[1]: *** [Makefile:331: dblat1] Error 1
```

But, Spack can be odd sometimes, so I grabbed OpenBLAS 0.3.29 and built it as:
```
make MAKE_NB_JOBS=0 ARCH=arm64 DYNAMIC_ARCH=1 DYNAMIC_OLDER=1 \
  TARGET=GENERIC DYNAMIC_ARCH=1 USE_LOCKING=1 USE_OPENMP=0 \
  USE_THREAD=0 RANLIB=ranlib NO_AVX512=1 all |& tee make.log
```
which is what Spack was building as. In this build, `CC=clang CXX=clang++ FC=gfortran-14` where `gfortran-14` is from Brew. 

Eventually my build failed with:
```
touch libopenblas-r0.3.29.a
/Applications/Xcode.app/Contents/Developer/usr/bin/make -C test all
/Users/mathomp4/.homebrew/brew/bin/gfortran-14 -O2 -Wall -frecursive -fno-optimize-sibling-calls -fPIC -fno-tree-vectorize -c sblat1.f  -o sblat1.o
/Users/mathomp4/.homebrew/brew/bin/gfortran-14 -O2 -Wall -frecursive -fno-optimize-sibling-calls  -fno-tree-vectorize  -o sblat1 sblat1.o ../libopenblas-r0.3.29.a -lgfortran -lgfortran -L/usr/local/lib  -lto_library -lSystem  /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/17/lib/darwin/libclang_rt.osx.a
ld: -lto_library library filename must be 'libLTO.dylib'
collect2: error: ld returned 1 exit status
make[1]: *** [sblat1] Error 1
make: *** [tests] Error 2
```

Now, I'm currently doing a reinstall of `gfortran-14` from source (as on this machine, I don't have admin access, so I have to install Brew in userspace), but is anyone else seeing this? Or have a solution?

> [@martin-frbg](https://github.com/martin-frbg?rgh-link-date=2025-04-03T22%3A47%3A25.000Z) I'll take a look on Monday when I'm back at work. So weird...
> 
> That said, I _think_ I was able to build OpenBLAS with pure gcc/gfortran and Spack (that is, avoiding Clang).

Note: Like @pbrady I was *NOT* able to build with pure gcc/gfortran:

```
==> Error: ProcessError: Command exited with status 2:
    '/Users/mathomp4/spack-mathomp4/opt/spack/darwin-m2/gmake-4.4.1-kz6bdpylf4e3yt4go4kiznkiouaf4wcb/bin/make' '-j6' '-s' 'CC=/Users/mathomp4/spack-mathomp4/opt/spack/darwin-m2/compiler-wrapper-1.0-cdasmd2yy77m4m6wp6mdpf72p6usoqcq/libexec/spack/gcc/gcc' 'FC=/Users/mathomp4/spack-mathomp4/opt/spack/darwin-m2/compiler-wrapper-1.0-cdasmd2yy77m4m6wp6mdpf72p6usoqcq/libexec/spack/gcc/gfortran' 'MAKE_NB_JOBS=0' 'ARCH=arm64' 'DYNAMIC_ARCH=1' 'DYNAMIC_OLDER=1' 'TARGET=GENERIC' 'DYNAMIC_ARCH=1' 'USE_LOCKING=1' 'USE_OPENMP=0' 'USE_THREAD=0' 'RANLIB=ranlib' 'NO_AVX512=1' 'all'

3 errors found in build log:
     6005       27 |     int i, j, info;
     6006          |               ^~~~
     6007    ld: warning: duplicate -rpath '/Users/mathomp4/.homebrew/brew/Cellar/gcc/14.2.0_1/lib/gcc/current/gcc/aarch64-apple-darwin23/14' ignored
     6008    ld: warning: duplicate -rpath '/Users/mathomp4/.homebrew/brew/Cellar/gcc/14.2.0_1/lib/gcc/current' ignored
     6009    ld: warning: ignoring duplicate libraries: '-lSystem', '-lemutls_w', '-lgfortran', '-lheapt_w'
     6010    ld: invalid r_symbolnum=25 in 'test_potrs.o'
  >> 6011    collect2: error: ld returned 1 exit status
  >> 6012    make[1]: *** [Makefile:78: openblas_utest] Error 1
  >> 6013    make: *** [Makefile:172: tests] Error 2

See build log for details:
  /var/folders/4t/b2c40c4149dfm599b9pc_7mr0000gp/T/mathomp4/spack-stage/spack-stage-openblas-0.3.29-xsdxb3g3jgvh2fpe2srz2o6m65m2uawb/spack-build-out.txt
```

ETA: It seems to die in the testing phase:
```
 Test of subprogram number  9            ZDSCAL
                                    ----- PASS -----

 Test of subprogram number 10            IZAMAX
                                    ----- PASS -----
test_kernel_regress.c: In function '__ctest_kernel_regress_skx_avx_run':
test_kernel_regress.c:27:15: warning: variable 'info' set but not used [-Wunused-but-set-variable]
   27 |     int i, j, info;
      |               ^~~~
ld: warning: duplicate -rpath '/Users/mathomp4/.homebrew/brew/Cellar/gcc/14.2.0_1/lib/gcc/current/gcc/aarch64-apple-darwin23/14' ignored
ld: warning: duplicate -rpath '/Users/mathomp4/.homebrew/brew/Cellar/gcc/14.2.0_1/lib/gcc/current' ignored
ld: warning: ignoring duplicate libraries: '-lSystem', '-lemutls_w', '-lgfortran', '-lheapt_w'
ld: invalid r_symbolnum=25 in 'test_potrs.o'
```

--- a/Makefile.system
+++ b/Makefile.system
@@ -435,6 +435,15 @@ ifeq (x$(XCVER), x 15)
 CCOMMON_OPT += -Wl,-ld_classic
 FCOMMON_OPT += -Wl,-ld_classic
 endif
+ifeq (x$(XCVER), x 16)
+ifeq ($(C_COMPILER), GCC)
+CCOMMON_OPT += -Wl,-ld_classic
+FCOMMON_OPT += -Wl,-ld_classic
+endif
+ifeq ($(F_COMPILER), GFORTRAN)
+override CEXTRALIB := $(filter-out(-lto_library, $(CEXTRALIB))) 
+endif
+endif
 endif
 
 ifneq (,$(findstring $(OSNAME), FreeBSD OpenBSD DragonFly))

--- a/ctest/Makefile
+++ b/ctest/Makefile
@@ -235,18 +235,18 @@ FLDFLAGS = $(FFLAGS:-fPIC=) $(LDFLAGS)
 ifeq ($(USE_OPENMP), 1)
 ifeq ($(F_COMPILER), GFORTRAN)
 ifeq ($(C_COMPILER), CLANG)
-CEXTRALIB += -lomp
+EXTRALIB += -lomp
 endif
 endif
 ifeq ($(F_COMPILER), NAG)
-CEXTRALIB = -lgomp
+EXTRALIB = -lgomp
 endif
 ifeq ($(F_COMPILER), IBM)
 ifeq ($(C_COMPILER), GCC)
-CEXTRALIB += -lgomp
+EXTRALIB += -lgomp
 endif
 ifeq ($(C_COMPILER), CLANG)
-CEXTRALIB += -lomp
+EXTRALIB += -lomp
 endif
 endif
 endif

--- a/test/Makefile
+++ b/test/Makefile
@@ -299,18 +299,18 @@ CLDFLAGS = $(CFLAGS) $(LDFLAGS)
 ifeq ($(USE_OPENMP), 1)
 ifeq ($(F_COMPILER), GFORTRAN)
 ifeq ($(C_COMPILER), CLANG)
-CEXTRALIB += -lomp
+EXTRALIB += -lomp
 endif
 endif
 ifeq ($(F_COMPILER), NAG)
-CEXTRALIB = -lgomp
+EXTRALIB = -lgomp
 endif
 ifeq ($(F_COMPILER), IBM)
 ifeq ($(C_COMPILER), GCC)
-CEXTRALIB += -lgomp
+EXTRALIB += -lgomp
 endif
 ifeq ($(C_COMPILER), CLANG)
-CEXTRALIB += -lomp
+EXTRALIB += -lomp
 endif
 endif
 endif

